File: objc/NSOperationObjCImageLoader/NSOperationObjCImageLoader/ImageDownloader.m
//
//  ImageDownloader.m
//  NSOperationObjCImageLoader
//
//  Created by zhen gong on 8/27/17.
//  Copyright © 2017 zhen gong. All rights reserved.
//

#import "ImageDownloader.h"
#import "PhotoRecord.h"
#import "NetworkManager.h"
#import <UIKit/UIKit.h>

@interface ImageDownloader()

@end

@implementation ImageDownloader

-(instancetype)initWithRecord:(PhotoRecord *)record{
    if(self = [super init]) {
        _photoRecord = record;
        self.name = record.name;
    }
    return self;
}

- (void)main {

    if(self.isCancelled) {
        return;
    }
    

//    NSData *data = [NSData dataWithContentsOfURL:_photoRecord.url];
//    if(data == nil || data.length == 0) {
//        _photoRecord.image = [UIImage imageNamed:@"Failed"];
//        _photoRecord.state = PhotoRecordStateFailed;
//        return;
//    }
//    _photoRecord.image = [UIImage imageWithData:data];
//    _photoRecord.state = PhotoRecordStateDownloaded;

    if([self.name isEqualToString:@"Volcan y saltos"]) {
        DLog(@"self.name:%@ state:%ld", self.name, (long)self.photoRecord.state);
        DLog(@"self.name:%@ url:%@", self.name, self.photoRecord.url);
    }
    
    [[NetworkManager sharedInstance] fetchPhotoDataFromURL:_photoRecord.url success:^(NSData *data) {
        if(self.isCancelled) {
            data = nil;
            return;
        }
        self.photoRecord.image = [UIImage imageWithData:data];
        self.photoRecord.state = PhotoRecordStateDownloaded;
        DLog(@"_photoRecord:%@ is successful.", self.photoRecord.name);
    } failure:^(NSError *error) {
        self.photoRecord.image = [UIImage imageNamed:@"Failed"];
        self.photoRecord.state = PhotoRecordStateFailed;
        DLog(@"_photoRecord:%@ is failed.", self.photoRecord.name);
    }];
    
}

@end

File: swift/NSOperationSwift/NSOperationSwift/NSOperation/ImageDownloader.swift
//
//  ImageDownloader.swift
//  NSOperationSwift
//
//  Created by zhen gong on 6/30/17.
//  Copyright © 2017 zhen gong. All rights reserved.
//

import UIKit

class ImageDownloader: Operation {

    //1 Add a constant reference to the PhotoRecord object related to the operation.
    let photoRecord : PhotoRecord
    
    //2 Create a designated initializer allowing the photo record to be passed in.
    init(photoRecord: PhotoRecord) {
        self.photoRecord = photoRecord
    }
    
    //3 main is the method you override in NSOperation subclasses to actually perform work.
    override func main(){
        
        //4 Check for cancellation before starting. Operations should regularly check if they have been cancelled before attempting long or intensive work.
        if self.isCancelled{
            return
        }
        //5 Download the image data.
        // let imageData = NSData(contentsOf: self.photoRecord.url as URL)
        // let imageData = Data(contentsOf: self.photoRecord.url)
        
        let imageData:Data? = try? Data(contentsOf: self.photoRecord.url)
        
        //6 Check again for cancellation.
        if self.isCancelled {
            return
        }
        
        //7 If there is data, create an image object and add it to the record, and move the state along. If there is no data, mark the record as failed and set the appropriate image.
        if (imageData?.count)! > 0 {
            self.photoRecord.image = UIImage(data: imageData! as Data)
            self.photoRecord.state = .Downloaded
        }
        else
        {
            self.photoRecord.state = .Failed
            self.photoRecord.image = UIImage(named: "Failed")
        }
        
    }

}
